### PR TITLE
support deepspeed  elastic 

### PR DESCRIPTION
**Pull request by meichangsu1**
### What changes were proposed in this pull request?

This pull request make dlrover support elatsic training when using deepspeed in ms-swift（we added checkpoint conversion logic before and after scaling operations and fixed some bug）

### Why are the changes needed?
With the current GPU shortage in our company, the elasticity feature can take advantage of idle resources, which is exactly what we need.

### Does this PR introduce any user-facing change?
perhaps

### How was this patch tested?
Validated in our uat  cluster  under GPU-limited conditions. Confirmed elastic scaling and checkpoint conversion logic worked as expected without regression.

### CODE REVIEW

File: dlrover/python/elastic_agent/torch/ckpt_saver.py [hunk 1]
```diff
@@ -423,19 +423,22 @@ def __init__(
...
             f"save_timeout={save_timeout}"
+            f"rank={rank}"
         )
         self.checkpoint_dir = checkpoint_dir
         self.local_shard_num = local_shard_num
         self.global_shard_num = global_shard_num
-        self._node_rank = env_utils.get_node_rank()
-        self._is_agent_rank_0 = self._node_rank == 0
+        self._node_rank = env_utils.get_rank()
+        self._rank=rank
```
Comment (BalaBalaYi) [inline dlrover/python/elastic_agent/torch/ckpt_saver.py]: local rank or global rank? need more annotations

File: dlrover/python/elastic_agent/torch/ckpt_saver.py [hunk 2]
```diff
@@ -502,15 +505,18 @@ def _factory():
...
                     if (
                         cls._saver_instance
                         and saver_thread
                         and saver_thread.is_alive()
                     ):
-                        logger.info(
-                            "The saver is already created, skip creating the saver."
-                        )
+
+                        # bugfix 因为checkpoint dir 是每次训练都会变化的
```
Comment (BalaBalaYi) [inline dlrover/python/elastic_agent/torch/ckpt_saver.py]: use english for all annotations

File: dlrover/python/master/elastic_training/rdzv_manager.py [hunk 3]
```diff
@@ -97,7 +98,16 @@ def get_max_nodes(self):
 
     def get_waiting_timeout(self):
         return self._rdzv_params.waiting_timeout
-
+    def get_ucp_ready(self):
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/rdzv_manager.py]: use @property for simple field return

File: dlrover/python/common/comm.py [hunk 4]
```diff
@@ -221,8 +221,12 @@ class GlobalStep(Message):
     timestamp: int = 0
     step: int = 1
     elapsed_time_per_step: float = 0.0
-
-
+@dataclass
+class UCPReady(Message):
```
Comment (BalaBalaYi) [inline dlrover/python/common/comm.py]: What does the UCP stand for? Unify Compute Platform? More explanation are needed.

File: dlrover/python/master/elastic_training/rdzv_manager.py [hunk 5]
```diff
@@ -166,6 +176,11 @@ def _check_rdzv_completed(self):
                 self.remove_alive_node(target_node)
 
         rdzv_completed = False
+        # 如果扩容前的universal checkpoint 没有就绪，下一轮rendezvou集会不能完成
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/rdzv_manager.py]: lint

File: docs/design/elastic.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
+# 弹性训练设计
```
Comment (BalaBalaYi) [inline docs/design/elastic.md]: By default a english version doc is needed. Besides a chinese version would be even better.

**Review by BalaBalaYi — CHANGES_REQUESTED** [review]
merge or rebase master please

File: dlrover/python/elastic_agent/torch/ckpt_saver.py (hunk 1 above)
Comment (meichangsu1) [reply dlrover/python/elastic_agent/torch/ckpt_saver.py]: need global rank to make `_dist_make_dir ` work as expected

File: dlrover/python/common/comm.py (hunk 4 above)
Comment (meichangsu1) [reply dlrover/python/common/comm.py]: universal checkpoint,相关文档也已经介绍

File: docs/design/elastic.md (hunk 6 above)
Comment (meichangsu1) [reply docs/design/elastic.md]: 已经改为英文

File: docs/tutorial/Elastic.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
+# Elastic
```
Comment (BalaBalaYi) [inline docs/tutorial/Elastic.md:1]:
```suggestion
# Elastic Training with MS-Swift
```

File: docs/design/elastic.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+### Universal Checkpoint (hereinafter referred to as UCP)
+
+```
+┌────────────────────┐
+│  Source Checkpoint │ (DP / TP / PP / SP)
+└──────────┬─────────┘
+           │
+           ▼
+   ┌──────────────────┐
+   │  UCP Atomic Files │  ← Each parameter stored independently (universal format)
```
Comment (BalaBalaYi) [inline docs/design/elastic.md]: the format of the illustration can be adjusted manually

File: docs/design/elastic.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
+# Elastic Training Design
```
Comment (BalaBalaYi) [inline docs/design/elastic.md]: Here, the document's name 'Elastic' and its content headings should ideally describe a scenario based on llm model training with PyTorch. 

It's important to clarify for readers that the "training elasticity" discussed here differs significantly from implementations based on TensorFlow's Parameter Server architecture.

File: dlrover/trainer/torch/flash_checkpoint/engine.py [hunk 10]
```diff
@@ -450,7 +475,9 @@ def get_global_shard_num(self):
         pass
 
     @abstractmethod
-    def save_to_memory(self, step, state_dict, paths: Dict[str, str]):
+    def save_to_memory(
+        self, step, state_dict, paths: Dict[str, str], blocking=False
```
Comment (BalaBalaYi) [inline dlrover/trainer/torch/flash_checkpoint/engine.py:473]: add annotation for new param: blocking

File: dlrover/trainer/torch/flash_checkpoint/engine.py [hunk 11]
```diff
@@ -194,6 +194,22 @@ def __init__(
...
+        self._rank = 0
+        self._group_rank = 0
+        self._world_size = 1
+        self._loader_group = None
+        self._saver_group = None
+        self._saving_ranks: Optional[List[int]] = None
+
+        self._init_sync_group(comm_backend)
+
+        self._rank = 0
```
Comment (BalaBalaYi) [inline dlrover/trainer/torch/flash_checkpoint/engine.py]: duplicated

File: dlrover/trainer/torch/flash_checkpoint/engine.py [hunk 12]
```diff
@@ -194,6 +194,22 @@ def __init__(
         self._local_rank = env_utils.get_local_rank()
         self._cached_step = -1
         self._restart_count = env_utils.get_torch_restart_count()
+        self._rank = 0
+        self._group_rank = 0
```
Comment (BalaBalaYi) [inline dlrover/trainer/torch/flash_checkpoint/engine.py:198]: use 'env_utils' to init(like 'self._local_rank')

File: dlrover/trainer/torch/flash_checkpoint/full_ckpt_engine.py [hunk 13]
```diff
@@ -133,11 +136,11 @@ def save_to_storage(self, step, state_dict, paths):
         """
         success = True
         if step > self._cached_step:
-            success = self.save_to_memory(step, state_dict, paths)
+            success = self.save_to_memory(step, state_dict, paths, blocking)
         # Only rank 0 persist the checkpoint to the storage.
         if dist.is_initialized():
             dist.barrier()
-        if success and self._rank == 0:
+        if success and self._local_rank == 0:
```
Comment (BalaBalaYi) [inline dlrover/trainer/torch/flash_checkpoint/full_ckpt_engine.py:143]: The modification here should adopt an extended logic. 

Otherwise, this change would cause all local_rank 0 processes to write repeatedly in a scenario where only global_rank 0 should write.

File: dlrover/trainer/torch/flash_checkpoint/engine.py [hunk 14]
```diff
@@ -408,7 +426,7 @@ def _restore_memory_from_replica(self):
             )
         dist.barrier()
 
-    def wait_latest_checkpoint(self, timeout=1800):
+    def wait_latest_checkpoint(self, timeout=1800, max_steps=None):
```
Comment (BalaBalaYi) [inline dlrover/trainer/torch/flash_checkpoint/engine.py:421]: add annotation for new param: max_steps

File: dlrover/trainer/torch/flash_checkpoint/engine.py [hunk 15]
```diff
@@ -348,13 +365,14 @@ def save_state_dict_to_memory(self, state_dict, conf: CheckpointConfig):
...
 
-        acquired = self._shm_lock.acquire(blocking=False)
+        acquired = self._shm_lock.acquire(blocking)
         logger.info(
             f"{self._rank}-{self._local_rank} acquired the lock of shared "
             f"memory: {acquired} for step: {conf.step}."
         )
         all_rank_ready = check_all_rank_ready(self._saver_group, acquired)
         if not all_rank_ready or not state_dict:
+            self.is_skip = True
```
Comment (BalaBalaYi) [inline dlrover/trainer/torch/flash_checkpoint/engine.py:367]: please explain this skip strategy？

File: dlrover/python/master/elastic_training/rdzv_manager.py [hunk 16]
```diff
@@ -88,6 +88,7 @@ def __init__(self):
         self._topology_sorter = DpTopologySorter()
         self._event_reporter = get_event_reporter()
         self.rendezvous_events: Dict[int, DurationSpan] = {}
+        self.ucp_ready = True
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/rdzv_manager.py]: Is necessary to discuss whether the central management of UCP should be placed under 'rdzv_manager'. This is because 'rdzv_manager' itself is designed to operate solely for rendezvous. The coupling of responsibilities here does not seem appropriate.

File: dlrover/trainer/torch/flash_checkpoint/engine.py (hunk 15 above)
Comment (meichangsu1) [reply dlrover/trainer/torch/flash_checkpoint/engine.py:367]: 这个标志用于标识当前step 是否被跳过，如果被跳过，是为了避免wait_latest_checkpoint这时候一直在等在最近的step结束

File: dlrover/trainer/torch/flash_checkpoint/full_ckpt_engine.py (hunk 13 above)
Comment (meichangsu1) [reply dlrover/trainer/torch/flash_checkpoint/full_ckpt_engine.py:143]: 每个process 的要保存的文件都存放到了ckpt_config中，每个process ckpt_config中存放的都是自己需要保存的文件，那么，当只有rank=0 保存文件，那么rank=1中ckpt_config记录的文件就会没保存，导致文件保存不完整

File: dlrover/trainer/torch/flash_checkpoint/engine.py (hunk 12 above)
Comment (meichangsu1) [reply dlrover/trainer/torch/flash_checkpoint/engine.py:198]: https://github.com/intelligent-machine-learning/dlrover/blob/4086ef1fede38ee2f5ca4972bcc7d4ac21f19e70/dlrover/trainer/torch/flash_checkpoint/engine.py#L227
我只是把_notify_agent_to_create_saver 这些逻辑移动到了_init_sync_group之后

File: dlrover/python/elastic_agent/torch/ckpt_saver.py [hunk 17]
```diff
@@ -1267,9 +1342,10 @@ class DdpCheckpointSaver(CommonDirCheckpointSaver):
     """Persist the checkpoint from CPU memory buffer into the storage."""
 
     def persist_to_storage(self, local_shard_id: int, ckpt_config):
-        if self._node_rank != 0:
-            logger.info("Skip and only rank 0 saves checkpoint in a DDP job.")
-            return
+        # In the DDP (Distributed Data Parallel) scenario, some files that needs to be saved by all ranks such as rng_state_1.pth.
```
Comment (samplise) [inline dlrover/python/elastic_agent/torch/ckpt_saver.py]: Delete those commented codes

File: dlrover/python/elastic_agent/torch/training.py [hunk 18]
```diff
@@ -1365,6 +1412,111 @@ def _save_ckpt_to_storage(self):
...
+                elapsed_time = time.time() - start_time
+
+                # Detect that a NEW checkpoint saving has actually started
+                if (
+                    start_saving_step is not None
+                    and last_success_step is not None
+                    and start_saving_step > last_success_step
+                ):
+                    target_step = start_saving_step
+                    seen_new_saving = True
```
Comment (samplise) [inline dlrover/python/elastic_agent/torch/training.py]: `seen_new_saving` -> `need_new_saving`

File: dlrover/python/master/elastic_training/rdzv_manager.py [hunk 19]
```diff
@@ -785,3 +833,21 @@ def _detect_stragglers(self):
...
         return stragglers
+
+
+def create_training_rdzv_manager() -> RendezvousManager:
+    """Factory to create the training rendezvous manager.
+
+    Use env var DLROVER_TRAINING_ELASTIC_MODE to select the implementation.
+    Supported values: "base", "ucp". Default is "ucp".
+    """
+    rdzv_type = os.getenv("DLROVER_TRAINING_ELASTIC_MODE", "base").lower()
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/rdzv_manager.py]: for master: use job args(defined in global_context.py) to control the logic here. 
also:  if there is new param for master
1) update the input params support in master.go
2) update the doc: argument.md 

for worker: use env(pass the env to worker when master create worker pod)

File: docs/design/elastic.md (hunk 9 above)
Comment (meichangsu1) [reply docs/design/elastic.md]: done
